### PR TITLE
Updates for Microsoft Azure AD authentication

### PR DESCRIPTION
**Pull request by kaaloo**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: 

Please see the table describing the parameters for the signin url here.

https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-protocols-oidc#send-the-sign-in-request

- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/openid-client/index.d.ts [hunk 1]
```diff
@@ -69,6 +70,14 @@ export class Client {
         readonly max_age?: number;
     }): Promise<TokenSet>;
 
+    oauthCallback(redirectUri: string, parameters: {}, checks?: {
+        readonly response_type?: string;
+        readonly state?: string;
+        readonly nonce?: string;
+        readonly code_verifier?: string;
+        readonly max_age?: number;
+    }): Promise<TokenSet>;
```
Comment (ulrichb) [inline types/openid-client/index.d.ts]: Here we should drop max_age and nonce. 

See 
> If you're using pure OAuth 2.0 then client.oauthCallback(redirectUri, parameters[, checks[, extras]]) is the OAuth 2.0 variant of this method, it has the same signature with the exception of checks only supporting code_verifier, state and response_type.

https://github.com/panva/node-openid-client/blob/master/docs/README.md#clientcallbackredirecturi-parameters-checks-extras

... and see also the code: https://github.com/panva/node-openid-client/blob/8d15d302f7fa53d953f032d07fc03b7f0a008285/lib/client.js#L488

File: types/openid-client/index.d.ts [hunk 2]
```diff
@@ -78,13 +87,17 @@ export class Client {
...
+    refresh(refreshToken: string | TokenSet): Promise<TokenSet>;
 }
 
 export class TokenSet {
     readonly access_token?: string;
     readonly token_type?: string;
     readonly id_token?: string;
     readonly refresh_token?: string;
+    readonly expires_at?: number;
+    readonly scope?: string;
```
Comment (ulrichb) [inline types/openid-client/index.d.ts]: `expires_at`: Okay.

But `scope` ?

See https://github.com/panva/node-openid-client/blob/master/docs/README.md#new-tokensetinput

**Review by ulrichb — CHANGES_REQUESTED** [review]
... could you also please add test cases  ...

File: types/openid-client/index.d.ts (hunk 1 above)
Comment (kaaloo) [reply types/openid-client/index.d.ts]: Thank you for the catch!

File: types/openid-client/index.d.ts (hunk 2 above)
Comment (kaaloo) [reply types/openid-client/index.d.ts]: `scope` is a Microsoft specific required parameter and is the rationale for this PR.  Without it I was unable to use the types as is.

https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-protocols-oidc#send-the-sign-in-request

> This request is similar to the first leg of the OAuth 2.0 authorization code flow, with these important distinctions:
> 
> The request must include the openid scope in the scope parameter.
> The response_type parameter must include id_token.
> The request must include the nonce parameter.

Comment (ulrichb) [reply types/openid-client/index.d.ts]: You linked a `.../oauth2/v2.0/authorize?..scope=..` URL => so the `AuthorizationUrlParameters` input applies (where `scope` already exists). I don't quite get why you are extending the `TokenSet` (which is primarily an output type).

... also please add test cases for your changes.

Thanks!

**Comment by kaaloo** [comment]
Closing this PR since node-openid client now carries its own types.

**Pull request closed without merging**